### DISCUSSION TITLE
Integrated tilemap editor

### DISCUSSION THREAD

**Discussion by AlexandreSi** [4ian/GDevelop#6809] [Enhancement ideas (and their technical discussions)]
Integrated tilemap objects were added to GDevelop end of July.

Here is a list of features to add to improve the usability of the object:
- ✅ Add 9-patch-like painting tool to speed up tilemap painting (Done in #6977)
- Define custom collision mask for each tile
- Rotate the tile before painting them [[Source]](https://forum.gdevelop.io/t/rotate-in-new-tilemap/60244)
- Y-sorting [[Source]](https://x.com/sanghendrix96/status/1819021180202025117)
- Auto tile  [[Source]](https://x.com/sanghendrix96/status/1819021180202025117)
- Animated tile [[Source]](https://forum.gdevelop.io/t/animatable-tilemap/60246)
- Add possibility to zoom in and out the tileset:
  - In the object editor to select easily tiles with hitboxes
  - In the instance properties panel to select easily the tile to use to paint
- When an instance of a tilemap is selected on the scene and the user is hovering one of the tile, display the tile coordinates somewhere
- When painting an instance on the scene, add shortcuts to flip a tile already set in the tilemap
- When painting an instance on the scene, find a way (alt click for instance on desktop) to select a tile already set in the tilemap to use it to paint new tiles
- Add a way to easily store data tile-wise (for instance the number of times it's been hit in order to change the tile accordingly)
  - An extension could do the job

**Comment by Silver-Streak** [comment]
> * When painting an instance on the scene, find a way (alt click for instance on desktop) to select a tile already set in the tilemap to use it to paint new tiles

For this one, I did some experimenting. Unless there's a way to do IDE extensions/interfaces soon, I'm not sure an extension could meet this need.

Can an extension be made to dynamically populate an array or structure with the grid x/y (and/or position) of each tile? yes.
Would that really accomplish this need? No not quite.

For more detail: 
1. As base variables have to be predeclared,  the extension would require the user to add some form of array/structure to the tilemap instance, and then pass that variable name as a parameter to the extension. Pretty messy and a bad UX.
2. Even for the use case mentioned above, not every tile is going to need the hit check/hit count/etc. The user would need to know the specific tile x/y or ID (or both) they care about, and then need to have some way to pass the specific info to an event (to check collision) and then back (to increase the "Hitcount" variable). 
   - Is some of this possible? Yes. Is an extension going to realistically help with this? no not particularly, other than maybe pre-building _some_ of the array children, but they'd still need to remember the exact tiles they care about in their expression.
4. The other main issue is that as far as I can tell, there's currently no way to check collision with specific tiles. You can check collision with the entire tilemap, but not Tile at GridX/Y or PositionX/Y. Maybe collision wouldn't be the right way to do that, but maybe a "Point is inside Tilemap Object GridX/y" or something like that?.

I think you could accomplish _some_ of this with separate hidden objects for collision, but there's still no great way to let people build out their data before the scene runs rather than via events, especially if they don't care about collision stuff, and I don't know that an extension would help with that, behavior, function, or otherwise.  I'm happy to be shown otherwise, but I couldn't figure out a great user friendly solution after poking at it over the past few days.

**Comment by joaovictordejesus** [comment]
Hello! Goodnight! I've been having a serious problem with the compilation for a week now. Every time I try to compile the project for Android, the message "Something went wrong :( Please try again). I'm finishing a project but I can't solve this problem. Please help me. Thank you!
![2024-08-02 014836](https://github.com/user-attachments/assets/e6978982-d59d-41d1-a7f4-e68a30fd73f3)

**Reply by Silver-Streak** [reply]
Hello! Sorry you're experiencing an issue. Please feel free to report a bug on the forums or as a new github issue. This discussion is specific to the tilemap editor. Your post will be hidden, sorry.

**Comment by Pyroqua** [comment]
Yay! some of my suggestions are being thought over.

**Comment by HelperWesley** [comment]
Hey. 👋

A few things I noticed that need to be changed/added to make it easier to use the tool are:

1. Middle mouse button and right click painting isn't ideal. It prevents users from using middle mouse to move the game scene while painting, and using right click to deselect or just do nothing would make more sense than using it to paint as well.

2. Can’t redo/undo properly. If you place a few tiles above your tilemap and try to undo them, the entire tilemap will move up. This will cause a host of issues for users if it's going to be constantly knocking their tilemap out of alignment with the rest of their game.

3. If you were set to painting before you press to preview the game, you can't deselect the tilemap without clicking back in to the game scene, meaning you'll always paint an extra tile by accident. You need to first realize you're still in paint mode, turn it off, and then you can click in to the scene to deselect your object. (I accidentally painted this way a dozen or more times.... 😅)

4. It would be great if we could use the condition to pick a specific tile without using to it's tile ID. This would allow users to pick any tile at a specific position instead of only 1 tile type at a specific position. It would allow for a lot more flexibility.

5. The tool seems to lag behind the mouse when placing tiles, we got a few comments on the YouTube channel with people saying it feels like a laggy tool.

**Reply by AlexandreSi** [reply]
1. Clearly a UX bug. Can be easily fixed to ignore those clicks or keep their functionalities.
2. I thought I had tested this, it might need a little bit of work but it should be doable.
3. I'm not sure how to fix this but I admit I ran into it many times. Maybe a clear hint that you are in painting mode?
4. Can you add an example, I'm not sure if you're talking about the tilemap or the tileset here
5. It is indeed laggy when you select which should have a collision mask (and that there are multiple selected tiles) but I don't recall seeing lags when painting on the scene. Do you have more details about this one?

**Reply by HelperWesley** [reply]
3. Maybe turning off the paint mode when you preview the game? (I'm not sure, just thinking out loud)

4. (Image included) The "value to compare" field, it prevents us from doing things like hovering the mouse over the tilemap and selecting any tile the mouse is currently over. It would be nice to either have another condition without the tile ID requirement, or make it option for this one.
![image](https://github.com/user-attachments/assets/75957513-aff3-4864-8318-6d630419704a)

**Reply by AlexandreSi** [reply]
3. I think that this would annoy some users that would like to preview many times while painting
4. Ok, so it could be an action that stores the tile (at a position/on the grid) in a variable

**Reply by HelperWesley** [reply]
3. Maybe, but since undo doesn't work properly right now, I think it's more inconvenient to undo and have it ruin your work than have to select the tile you want to paint each time you preview the game. ⚖

4. Maybe? I just want to be able to pick a tile at any position, so I can dynamically change the tilemap without needing to know the tile ID.

**Reply by AlexandreSi** [reply]
3. @LuniMoon is there something that you can pick from your magical UX hat?
4. I realized it's currently doable:
  - Use the action "Tile at position"
  - for the tile id, use the expression "TileIdAtPosition"
    <img width="885" alt="image" src="https://github.com/user-attachments/assets/8f390e38-177c-42e3-ac68-0db08f94da9e">

**Reply by HelperWesley** [reply]
3. Luni 🙏

4. Oh, that might be what I needed. 🤔 I'll report back if I run in to anything I can't do with this. 👍

**Reply by HelperWesley** [reply]
5. Here's a video showing the tiles not following the mouse as closely as a sprite object does. (I've heard from users on YouTube that this is maybe more dramatic on slower computers.)

https://github.com/user-attachments/assets/ba4d6896-81bc-4a15-8051-042e89332fab

**Reply by AlexandreSi** [reply]
5. Thanks for the example. I removed the FPS limitation that gave that feeling. It will be available in the next version.
